lib/actions/build.coffee

```diff
@@ -0,0 +1,224 @@
+
+	# Now need to check if the user wants to connect over TLS
+	# to the host
+	connectOpts['ca'] = options.ca if options.ca?
+	connectOpts['cert'] = options.cert if options.cert?
+	connectOpts['key'] = options.key if options.key?
+
+	return connectOpts
+
+module.exports =
+	signature: 'build [context]'
+	description: 'Build a container locally'
+	permission: 'user'
+	options: [
+		{
+			signature: 'docker'
+			parameter: 'docker'
+			description: 'Path to a local docker socket'
+			alias: 'P'
+		},
+		{
+			signature: 'arch'
+			parameter: 'arch'
+			description: 'The architecture to build for'
+			alias: 'A'
+		},
+		{
+			signature: 'deviceType'
+			parameter: 'deviceType'
+			description: 'The type of device this build is for'
+			alias: 'd'
+		},
+		{
+			signature: 'context'
+			parameter: 'context'
+			description: "The build context, which contains the build files. Defaults to '.'"
+			alias: 'c'
+		},
+		{
+			signature: 'application'
+			parameter: 'application'
+			description: 'The target resin.io application this build is for'
+			alias: 'a'
+		},
+		{
+			signature: 'tag'
+			parameter: 'tag'
+			description: 'The alias to the generated image'
+			alias: 't'
+		},
+		{
+			signature: 'nocache'
+			description: "Don't use cache when building"
+			boolean: true
+		},
+		{
+			signature: 'dockerHost'
+			parameter: 'dockerHost'
+			description: 'The address of the host containing the docker daemon'
+			alias: 'h'
+		},
+		{
+			signature: 'dockerPort'
+			parameter: 'dockerPort'
+			description: 'The port on which the host docker daemon is listening'
+			alias: 'p'
+		},
+		{
+			signature: 'ca'
+			parameter: 'ca'
+			description: 'TLS Certificate authority file'
+		},
+		{
+			signature: 'cert'
+			parameter: 'cert'
+			description: 'TLS Certificate file'
+		},
+		{
+			signature: 'key'
+			parameter: 'key'
+			description: 'TLS Key file'
+		}
+	]
+	action: (params, options, done) ->
+		# The default build context is the current directory
+		params.context ?= '.'
+
+		# Tar up the directory, ready for the build stream
+		tarDirectory(params.context)
+		.then (tarStream) ->
+			hooks =
+					buildSuccess: (image, layers) ->
+						if options.tag?
+							console.log("Tagging image as #{options.tag}")
+						done()
+					buildFailure: (err) ->
+						done("Build error: #{err.message}")
+					buildStream: (stream) ->
+						info = getBundleInfo(options)
+						# info is either a promise, or is undefined
+						if info?
+							info
+							.catch (err) -> done("Error: #{err.message}")
+							.then ([arch, deviceType]) ->
+								# Perform type resolution on the project
+								bundle = new resolve.Bundle(tarStream, deviceType, arch)
+								resolve.resolveBundle(bundle, resolve.getDefaultResolvers())
+							.then (resolved) ->
+								console.log("Building #{resolved.projectType} project")
+								# Send the resolved tar stream to the docker daemon
+								resolved.tarStream.pipe(stream)
+							.catch(done)
```

Comment:
Ah I was having some issues with error propagation yesterday, this is an artifact of that. Will remove.